Fix: TimberExceptionLogging gives false positive for guarded variables

jamie10x opened the pull request:
The `TimberExceptionLogging` lint check was incorrectly flagging valid code where a message variable was guarded by an `isNotEmpty()` check.

This was caused by two issues in the detector:
1. The check was not flow-sensitive and couldn't determine that the variable was guaranteed to be non-empty inside the `if` block. It defaulted to flagging any unevaluated variable.
2. The logic for identifying arguments was based on position, causing it to fail when Kotlin's named or reordered arguments were used.

This commit resolves both issues by:
- Making the check more conservative: It now only flags issues it can prove are incorrect (e.g., literal empty strings or redundant `e.getMessage()` calls), avoiding false positives on unevaluated variables.
- Robustly identifying throwable and message arguments by their type, making the check work correctly regardless of argument order.

A new test case has been added to verify the fix and prevent regressions.

Fixes #331

jamie10x committed:
Fix: TimberExceptionLogging gives false positive for guarded variables

The `TimberExceptionLogging` lint check was incorrectly flagging valid code where a message variable was guarded by an `isNotEmpty()` check.

This was caused by two issues in the detector:
1. The check was not flow-sensitive and couldn't determine that the variable was guaranteed to be non-empty inside the `if` block. It defaulted to flagging any unevaluated variable.
2. The logic for identifying arguments was based on position, causing it to fail when Kotlin's named or reordered arguments were used.

This commit resolves both issues by:
- Making the check more conservative: It now only flags issues it can prove are incorrect (e.g., literal empty strings or redundant `e.getMessage()` calls), avoiding false positives on unevaluated variables.
- Robustly identifying throwable and message arguments by their type, making the check work correctly regardless of argument order.

A new test case has been added to verify the fix and prevent regressions.

Fixes #331

JakeWharton commented on timber-lint/src/test/java/timber/lint/WrongTimberUsageDetectorTest.kt:
```diff
@@ -1425,4 +1425,38 @@ class WrongTimberUsageDetectorTest {
         .expectClean()
   }
+
+    @Test fun exceptionLoggingWithGuardedVariable() {
```
The indent here is wrong. Hopefully Spotless/ktlint will flag. GitHub won't let me run the checks so probably gotta kick it with a rebase or a merge.

JakeWharton commented:
Looks like one test is failing on detection of `null` message.

jamie10x committed:
Fix: TimberExceptionLogging gives false positive for guarded variables

The `TimberExceptionLogging` lint check was incorrectly flagging valid code where a message variable was guarded by an `isNotEmpty()` check.

This was caused by two issues in the detector:
1. The check was not flow-sensitive and couldn't determine that the variable was guaranteed to be non-empty inside the `if` block. It defaulted to flagging any unevaluated variable.
2. The logic for identifying arguments was based on position, causing it to fail when Kotlin's named or reordered arguments were used.

This commit resolves both issues by:
- Making the check more conservative: It now only flags issues it can prove are incorrect (e.g., literal empty strings or redundant `e.getMessage()` calls), avoiding false positives on unevaluated variables.
- Robustly identifying throwable and message arguments by their type, making the check work correctly regardless of argument order.

A new test case has been added to verify the fix and prevent regressions.

Fixes #331

jamie10x committed:
Merge branch 'JakeWharton:trunk' into fix/331-exception-logging-false-positive

jamie10x committed:
Fix failing tests from PR review

This commit addresses the feedback from the PR review:

1. Corrected the logic in `checkExceptionLogging` to properly detect literal `null` message arguments. This fixes the failing `exceptionLoggingUsingNullMessage` test.

2. Updated the lint baseline to include the new test case, resolving the `LintImplTrimIndent` CI failure.

3. Corrected manual formatting/indentation of the new test.

jamie10x commented:
I've pushed up a new commit that should resolve the CI failure. The logic now correctly handles the `null` message case, and I've updated the lint baseline to account for the new test.

All 41 tests are passing locally on my end.

JakeWharton commented:
Thanks!
